pkg/chunked/storage_linux.go

```diff
@@ -235,30 +242,98 @@ func copyFileFromOtherLayer(file internal.FileMetadata, source string, otherFile
 	return true, dstFile, written, err
 }
 
+// canDedupMetadataWithHardLink says whether it is possible to deduplicate file with otherFile.
+// It checks that the two files have the same UID, GID, file mode and xattrs.
+func canDedupMetadataWithHardLink(file *internal.FileMetadata, otherFile *internal.FileMetadata) bool {
+	if file.UID != otherFile.UID {
+		return false
+	}
+	if file.GID != otherFile.GID {
+		return false
+	}
+	if file.Mode != otherFile.Mode {
+		return false
+	}
+	if !reflect.DeepEqual(file.Xattrs, otherFile.Xattrs) {
+		return false
+	}
+	return true
+}
+
+// canDedupFileWithHardLink checks if the specified file can be deduplicated by an
+// open file, given its descriptor and stat data.
+func canDedupFileWithHardLink(file *internal.FileMetadata, fd int, s os.FileInfo) bool {
+	st, ok := s.Sys().(*syscall.Stat_t)
+	if !ok {
+		return false
+	}
+
+	path := fmt.Sprintf("/proc/self/fd/%d", fd)
+
+	listXattrs, err := system.Llistxattr(path)
+	if err != nil {
+		return false
+	}
+
+	xattrsToIgnore := map[string]interface{}{
+		"security.selinux": true,
```

Comment:
Why is this {necessary, safe}?

Reply:
I think @giuseppe is thinking that containers use a volume mount and do not rely on the label. But as @mtrmac this might actually break things, since a container might not be able to use the SELinux label in the lower.  The problem is going to be figuring out which labels can and can not be used.  If the label came from container storage it should be fine, and the check would not be necessary.  If the link came from the host system like in /usr, it probably would work.  If the link came from /var or other locations on the host system, it probably will not be readable within the confined container.

Reply:
IIRC host reuse only uses `/usr`, so this might be fine — but having the full thinking documented in the code seems like something pretty likely to be useful 3–5 years from now, both to find out what the original motivation was, and so that future programmers and PRs that touch the same code notice that there’s something to think about.

This PR comment thread (hopefully we settle on a full definite story) can help recording what we were thinking, but a comment directly in code would help draw attention to the issue.

Reply:
The question would be is can a user turn this off if it does cause problems?

Reply:
“Problems?”

This seems to be to be the kind of thing where expert programmers should either determine that it works and that it is safe, or not have that feature; I find it unintuitive to treat this like performance tuning knob where there are many legitimate options and the best one is based on circumstances.

Reply:
… and if we have no idea whether this is safe and works, why not just match the SELinux label exactly, and reuse files only in the case where we do have a good reason to think it is safe and works?

---

@giuseppe can you explain what you saw and how this works, please?

Reply:
Well the content will never match if you are searching through /usr, versus /var/lib/containers/storage.

Reply:
That’s a fair point. So what’s the property that we want (even if this _were_ a manual user option, what information would the user need to make a decision?), and the right way to check? A run-time read of the policy to see whether a specific operation is allowed? A build-time provided list of SELinux types that should be accepted, and compiling that list into the binary?

Looking at an old `/usr/bin` or  `/usr/libexec`, not just a `bin_t` or `lib_t` or `usr_t`; there are quite a few other types. What property are we looking for, and do they have that property?

Reply:
$ sesearch -A -s container_t -t usr_t -c file -p read
allow domain base_ro_file_type:file { getattr ioctl lock open read };
allow svirt_sandbox_domain exec_type:file { entrypoint execute execute_no_trans getattr ioctl lock map open read };

$ seinfo -abase_ro_file_type -x

Type Attributes: 1
   attribute base_ro_file_type;
	bin_t
	boot_t
	etc_runtime_t
	etc_t
	lib_t
	shell_exec_t
	src_t
	system_conf_t
	system_db_t
	textrel_shlib_t
	usr_t

Reply:
We could just hard code this list into opencontainers/selinux and then have storage use it.

Reply:
Any other types we ignore.

Reply:
Yup, that looks like a consistent story to (ignorant) me. Is the list of `base_ro_file_type` types relatively static? Or would we want to extract it at RPM build time? (OTOH that could potentially unexpectedly change the list on simple rebuilds.)

Reply:
It has been static for years.